mem-store.ts

```diff
@@ -106,18 +108,97 @@ function mem_store(options: any) {
         if (shouldMerge) {
           mement = Object.assign(prev || {}, mement)
         }
-        prev = entmap[base][name][mement.id] = mement
 
-        seneca.log.debug(function () {
-          return [
-            'save/' + (create ? 'insert' : 'update'),
-            ent.canon$({ string: 1 }),
-            mement,
-            desc,
-          ]
+        return upsertIfRequested(msg, function (err: Error, ctx: UpsertIfRequestedContext) {
+          if (err) {
+            return reply(err)
+          }
+
+          const { did_upsert } = ctx
+
+          if (did_upsert) {
+            const { out } = ctx
+            return reply(null, out)
+          }
+
+
+          prev = entmap[base][name][mement.id] = mement
+
+          seneca.log.debug(function () {
+            return [
+              'save/' + (isNewEntityBeingCreated(msg) ? 'insert' : 'update'),
+              ent.canon$({ string: 1 }),
+              mement,
+              desc,
+            ]
+          })
+
+          return reply(null, ent.make$(prev))
         })
 
-        reply(null, ent.make$(prev))
+
+        type UpsertIfRequestedContext = { did_upsert: boolean, out?: any }
+
+        type UpsertIfRequestedCallback = (
+          err:     Error | null,
+          result?: UpsertIfRequestedContext 
+        ) => any
+
+        function upsertIfRequested(msg: any, reply: UpsertIfRequestedCallback) {
+          // This is the query passed to the .save$ method.
+          //
+          const query_for_save = msg.q
+
+
+          if (isNewEntityBeingCreated(msg) && Array.isArray(query_for_save.upsert$)) {
+            const upsert_on = Common.clean(query_for_save.upsert$)
+
+
+            if (upsert_on.length > 0) {
+              if (!(base in entmap)) {
```

Comment:
While technically this check is unnecessary due to: https://github.com/senecajs/seneca-mem-store/blob/ed2dec365ef55e6b589de7a80f69ff386a212007/mem-store.ts#L89

- I think it is important to include it here. The reason is that in this manner `upsertIfRequested` is decoupled from the knowledge about the inner workings of the `do_save` function.

The only thing `upsertIfRequested` is concerned about, is that if there is no store for the requested entity, then there can be no match - and as a result - no update.